pandas/core/internals/blocks.py

```diff
@@ -1969,22 +1970,19 @@ def shift(self, periods, axis=0):
 
     def where(self, other, cond, align=True, errors='raise',
               try_cast=False, axis=0, transpose=False):
-        # Extract the underlying arrays.
-        if isinstance(other, (ABCIndexClass, ABCSeries)):
-            other = other.array
-
-        elif isinstance(other, ABCDataFrame):
+        if isinstance(other, ABCDataFrame):
             # ExtensionArrays are 1-D, so if we get here then
             # `other` should be a DataFrame with a single column.
             assert other.shape[1] == 1
-            other = other.iloc[:, 0].array
+            other = other.iloc[:, 0]
+
+        other = extract_array(other, extract_numpy=True)
```

Comment:
is extract_numpy ever not True?

Reply:
Yes, when we need the EA-like behavior, e.g. in reshape: https://github.com/pandas-dev/pandas/pull/24227/files#diff-fef81b7e498e469973b2da18d19ff6f3R437